### PR TITLE
bug 1870728: pkg/asset/ignition/bootstrap: Display warning if certificates are expired

### PR DESCRIPTION
**Pull request by jhixson74**
When loading bootstrap.ign, check if there are expired certificates and display a warning if so.

https://bugzilla.redhat.com/show_bug.cgi?id=1870728

### CODE REVIEW

**Comment by rvanderp3** [comment]
> When loading bootstrap.ign, check if there are expired certificates and display a warning if so. I'm not coupling this to the bug until I get verification this is going the right direction ;-)
> 
> https://bugzilla.redhat.com/show_bug.cgi?id=1870728

Yes, this is what I was looking for in the BZ but there may be other thoughts on the implementation.  I like this a lot more than the solution I had put together in the PR attached to the BZ.

**Review by mtnbikenc — APPROVED** [review]

**Comment by mtnbikenc** [comment]
Just noticed a small golint error that needs to be addressed.

**Review by mtnbikenc — CHANGES_REQUESTED** [review]
Just noticed a small golint issue that needs to be addressed.
/lgtm cancel

File: pkg/asset/ignition/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -557,5 +561,41 @@ func (a *Bootstrap) Load(f asset.FileFetcher) (found bool, err error) {
...
 	a.File, a.Config = file, config
+	warnIfCertificatesExpired(a.Config)
 	return true, nil
 }
+
+// warnIfCertificatesExpired checks for expired certificates and warns if so
+func warnIfCertificatesExpired(config *igntypes.Config) {
+	expiredCerts := 0
+	for _, file := range config.Storage.Files {
+		if strings.HasSuffix(file.Path, ".crt") && file.Contents.Source != nil {
```
Comment (abhinavdahiya) [inline pkg/asset/ignition/bootstrap/bootstrap.go:572]: https://golang.org/pkg/path/filepath/#Ext

File: pkg/asset/ignition/bootstrap/bootstrap.go [hunk 2]
```diff
@@ -557,5 +561,41 @@ func (a *Bootstrap) Load(f asset.FileFetcher) (found bool, err error) {
...
+			// ^data:text/plain;charset=utf-8;base64,LS0.+==
+			i := strings.Index(*file.Contents.Source, ",LS0")
+			if i == -1 {
+				continue
+			}
```
Comment (abhinavdahiya) [inline pkg/asset/ignition/bootstrap/bootstrap.go:577]: https://github.com/vincent-petithory/dataurl I think ignition uses this..

File: pkg/asset/ignition/bootstrap/bootstrap.go [hunk 3]
```diff
@@ -557,5 +561,41 @@ func (a *Bootstrap) Load(f asset.FileFetcher) (found bool, err error) {
...
+			block, _ := pem.Decode(data)
+			if block != nil {
+				cert, err := x509.ParseCertificate(block.Bytes)
+				if err == nil {
+					if time.Now().UTC().After(cert.NotAfter) {
+						logrus.Warnf("Bootstrap Ignition-Config Certificate %s expired at %s.", path.Base(file.Path), cert.NotAfter.Format(time.RFC3339))
+						expiredCerts++
+					}
+				} else {
+					logrus.Debugf("Unable to parse certificate. %s", err.Error())
```
Comment (abhinavdahiya) [inline pkg/asset/ignition/bootstrap/bootstrap.go:592]: some information on which certificate would be helpful in the logs

File: pkg/asset/ignition/bootstrap/bootstrap.go [hunk 4]
```diff
@@ -557,5 +561,41 @@ func (a *Bootstrap) Load(f asset.FileFetcher) (found bool, err error) {
...
+			i := strings.Index(*file.Contents.Source, ",LS0")
+			if i == -1 {
+				continue
+			}
+			data, err := base64.StdEncoding.DecodeString((*file.Contents.Source)[i+1:])
+			if err != nil {
+				logrus.Debugf("Unable to decode certificate. %s", err.Error())
+				continue
+			}
+			block, _ := pem.Decode(data)
```
Comment (abhinavdahiya) [inline pkg/asset/ignition/bootstrap/bootstrap.go:583]: I think this assumes there will only be one certificate in the file??

File: pkg/asset/ignition/bootstrap/bootstrap.go (hunk 1 above)
Comment (jhixson74) [reply pkg/asset/ignition/bootstrap/bootstrap.go]: Fixed.

File: pkg/asset/ignition/bootstrap/bootstrap.go (hunk 2 above)
Comment (jhixson74) [reply pkg/asset/ignition/bootstrap/bootstrap.go]: Fixed.

File: pkg/asset/ignition/bootstrap/bootstrap.go (hunk 4 above)
Comment (jhixson74) [reply pkg/asset/ignition/bootstrap/bootstrap.go]: Is there ever a case where there is a certificate chain in bootstrap.ign? If not, this works as intended. Every certificate in bootstrap.ign is checked for expiration.

File: pkg/asset/ignition/bootstrap/bootstrap.go (hunk 3 above)
Comment (jhixson74) [reply pkg/asset/ignition/bootstrap/bootstrap.go]: Fixed.

**Pull request merged**